### ISSUE TITLE
Move Openverse API and catalog to `openverse.org` subdomains

### ISSUE THREAD

**Issue by sarayourfriend** [WordPress/openverse#2037]
## Summary

Move the infrastructure which currently exists on `openverse.engineering` to `openverse.org`.

## Description

<details>
<summary>Pseudo-project proposal (collapsed to preference the issue and document list)</summary>

We currently pay for two Cloudflare accounts, one for `openverse.engineering` and `openverse.org`. If we move the API and catalog to live on subdomains of `openverse.org`, we should be able to change our `openverse.engineering` account to a free one, saving 200 USD a month that can go towards e.g., Plausible.

In the initial discussion we had about this we only talked about moving the API. However, because Airflow is behind Cloudflare Access, we may need to move it as well. I've tried to understand whether that is the case based on the Cloudflare pricing and it seems like Cloudflare Access might be free under our current usage but it isn't clear to me whether that's free _for specific paid account types_ or free for _any_ account type.

The end result of this project should be that `openverse.engineering` and its usage should be entirely covered by a free Cloudflare account. The planning for this project must consider the various features we will need that will cause contingencies here:

1. Do we need to route redirects through Cloudflare, or could we do it AWS side with a small redirect service? If we do it via Cloudflare, we might need individual page rules for each domain we are redirecting. For the API there are at least two we need to redirect: `api.openverse.engineering` and `api-production.openverse.engineering`. Do we need to redirect staging (`api-staging`)? What about the legacy staging subdomain, `api-dev`? We also continue to redirect `search.openverse.engineering`, our legacy frontend domain(s). Cloudflare free only supports three page rules. We currently have 3 redirects already (for the frontend). We would need at minimum one more for the API (`api.openverse.engineering`) but to match our existing redirect philosophy with the frontend, we should redirect `api`, `api-production` and `api-staging`. That would lead to 6 minimum page rules if we used page rules for the redirects. The other existing page rules are caching related and would be moved into `openverse.org`, so they do not need to count towards our total page rule utilisation.
2. Can the catalog stay on `openverse.engineering` if it is a free account? Namely, as discussed above, do free Cloudflare accounts support Cloudflare Access? If we manage Cloudflare Access on two different Cloudflare zones, the access terraform module will need to be updated or potentially completely re-written to accommodate multiple Cloudflare zones.

Additional considerations:

- The jumphost is on `openverse.engineering`. Should is stay there?
- If we need to configure more redirects than page rules allow on a free account, can a small ECS Nginx or caddy instance be configured to properly handle every redirect we need? Is it cheaper to host an ECS service for this or a small EC2 instance or can it be managed entirely through AWS ALB listener rules instead?
- Special attention must be paid to timing the transitions and announcements on Make. Should a notice be sent to registered API users?
- Updates to API domains will be necessary in many parts of our stack, including documentation and the frontend configuration. Frontend test tapes will need to be updated for the new domain. Are there other instances of such fixtures that we would need to update?

A task list was written by @zackkrida and @dhruvkb before. It is listed below in a collapsed element as I think it should be referenced with a grain of salt. It demonstrates the overall picture well, especially for communications, but hides what I think are probably the most complex parts of this (namely the individual infrastructure steps we need to take for the first task).
</details>

<details>
<summary>The task list created during the original, internal discussion and proposal of this project.</summary>

- Infrastructure PR to point the new domains (can be done in advance) and allow traffic from openverse.org to the load balancer API service
- Add Cloudflare page cache rules for the API in openverse.org
- Follow-up infrastructure PR to switch the old domains to 301 redirects
- Write a make post
- Get a list of all unique registered email addresses (pretty easy with the jumphost and a DB query) and draft an email
- Publish the make post and send the email
- Update references to https://api(-staging)?.openverse.engineering in the documentation
- On launch day: merge the infra PR to turn the old domains into redirects
- On launch day: Downgrade the Cloudflare account for openverse.engineering
- On launch day: Announce completion of the redirect on the make post and in make slack
</details>

## Documents

Because this project is relatively clear in its motivations and requirements, we will skip the project proposal. This project thread's description will serve as a general project proposal.

- [x] https://github.com/WordPress/openverse/issues/2038

## Issues

Issues are mostly in the infrastructure repository, organised into the following four milestones.

- [x] [Preliminary work](https://github.com/WordPress/openverse-infrastructure/milestone/4) (Completed :tada:)
- [x] [Airflow work](https://github.com/WordPress/openverse-infrastructure/milestone/6) (Completed :tada:)
- [x] [API work](https://github.com/WordPress/openverse-infrastructure/milestone/5) (In progress)
- [ ] [Finalisation work](https://github.com/WordPress/openverse-infrastructure/milestone/7) (Blocked)

Preliminary work blocks the API and Airflow specific work. API and Airflow work can happen in parallel. Finalisation work is blocked by everything else.

Additionally, these four issues are in the monorepo related to this:

- https://github.com/WordPress/openverse/issues/3741
- https://github.com/WordPress/openverse/issues/3742
- https://github.com/WordPress/openverse/issues/3743
- https://github.com/WordPress/openverse/issues/3821

**Comment by sarayourfriend** [comment]
Here are my answers to the questions I raised in the issue description:

> Do we need to route redirects through Cloudflare, or could we do it AWS side with a small redirect service?

If there are not sufficient Cloudflare rules in the free tier to redirect everything that needs it, we can use [AWS load balancer rules](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/listener-update-rules.html). I don't have a strong preference either way, maybe with a slight lean towards Cloudflare because the rules are more flexible, and technically they're at the edge, and would perform slightly better than load balancer rules. But those benefits are _super_ negligible, so if there is any hassle at all making it work in Cloudflare, AWS LB rules are a perfectly good fallback that will 100% work and will not cost us anything.

> Can the catalog stay on openverse.engineering if it is a free account?

No reason to do this. Let's move everything over and call it a day.

> The jumphost is on openverse.engineering. Should is stay there?

This is no longer true. The new SSH bastion is on `openverse.org`

> can a small ECS Nginx or caddy instance be configured to properly handle every redirect we need?

**NO**. We will use AWS load balancer rules. _We must avoid any solutions that introduce new services!_

> Should a notice be sent to registered API users?

Yes.

> Updates to API domains will be necessary in many parts of our stack, including documentation and the frontend configuration. Frontend test tapes will need to be updated for the new domain. Are there other instances of such fixtures that we would need to update?

The implementation plan will cover this.

**Comment by sarayourfriend** [comment]
Implementation plan is approved and merged (thanks @dhruvkb and @AetherUnbound for the careful review).

I've created all the issues for this project. They are primarily in the private infrastructure repository, almost all the work will happen there. The milestones are as follows:

- [Preliminary work](https://github.com/WordPress/openverse-infrastructure/milestone/4)
- [Airflow work](https://github.com/WordPress/openverse-infrastructure/milestone/6)
- [API work](https://github.com/WordPress/openverse-infrastructure/milestone/5)
- [Finalisation work](https://github.com/WordPress/openverse-infrastructure/milestone/7)

Preliminary work blocks the API and Airflow specific work. API and Airflow work can happen in parallel. Finalisation work is blocked by everything else.

Additionally, these three issues are in the monorepo related to this:

- https://github.com/WordPress/openverse/issues/3741
- https://github.com/WordPress/openverse/issues/3742
- https://github.com/WordPress/openverse/issues/3743

**Comment by AetherUnbound** [comment]
Thanks Sara for all this careful planning! I've gone ahead and moved all of the non-blocked tickets in the first preliminary work milestone into our TODOs.

**Comment by sarayourfriend** [comment]
Thanks for mentioning the blocked issues in the preliminary work milestone. Those are actually _finalisation_ work, I'd just forgotten to move them! They're in the correct milestone now :+1:

**Comment by sarayourfriend** [comment]
Preliminary work for this is finished now. I will start working on preparing the Airflow migration now.

**Comment by zackkrida** [comment]
I wanted to make a note here concerning CloudFlare and some of the currently-manual configuration for dealing with bots. On the frontend we now have "super bot fight mode" enabled, which automatically blocks all traffic from known and likely bad bots, while allowing ["verified" web crawlers](https://radar.cloudflare.com/traffic/verified-bots) like Internet Archive, Google, Bing, etc. to access the frontend. 

After moving to the openverse.org domain, we probably want to create Web Access Firewall rules to [skip](https://developers.cloudflare.com/waf/custom-rules/skip/) "super bot mode" rules for the API. I would guess at least that our API users programmatically accessing the API would be marked as bots by Cloudflare and blocked by these rules.

Specifically, our WAF rules need to skip the "http_request_sbfm" (sbfm = super bot fight mode) [request phase](https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs/resources/ruleset#phase). I think the whole rule would (roughly) look like this:

```tf
# https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs/resources/ruleset
resource "cloudflare_ruleset" "skip_sbfm_for_api" {
  zone_id = var.cloudflare_zone_id
  name    = "Skip Super Bot Fight Mode for the API"
  kind    = "zone"
  phase   = "http_request_sbfm"

  rules {
    action     = "skip"
    expression = "(http.request.uri.path matches \"^/api\")"
    description = "Skip Super Bot Fight Mode for requests to the API"
  }
}
```

**Comment by sarayourfriend** [comment]
@zackkrida Can you please add this note to the issue for moving all existing Cloudflare rules from the .engineering zone to the .org zone? It's unlikely the implementer of that issue would refer to this project thread before starting it, so it's more reliable to share that information there. The issue should have all the information, ideally.

https://github.com/WordPress/openverse-infrastructure/issues/777

**Comment by zackkrida** [comment]
@sarayourfriend I've updated my [comment](https://github.com/WordPress/openverse/issues/2037#issuecomment-1991856437) to match the hostname correctly. I've also manually added this rule to Cloudflare _now_, so no action needs to be taken in https://github.com/WordPress/openverse-infrastructure/issues/777. 

> Is the main issue needing to make sure that the existing rule for the frontend in the .org zone not accidentally cause an issue to the API?

Yes, the goal is so that Super Bot Fight Mode doesn't block programmatic API traffic once it's moved over to openverse.org. We do not need to bypass this for Kibana or Airflow.

**Comment by sarayourfriend** [comment]
> I've also manually added this rule to Cloudflare now

As in, added to the .org zone? Will you open a PR to add it to the cloudflare root module of the infrastructure repository? Ideally we move away from the practice of manually defining rules in Cloudflare without reviews or the chance to document in comments on them, particularly for things we expect to exist for a long time or indefinitely.

**Comment by zackkrida** [comment]
@sarayourfriend agreed on no longer manually defining these in the cf ui,I mentioned in my [edit](https://github.com/WordPress/openverse/issues/2037#issuecomment-1991856437) that we should move this firewall rule with the others.

**Comment by sarayourfriend** [comment]
Okay, thanks.

**Comment by sarayourfriend** [comment]
Yesterday I got up the PR to deploy Airflow with Ansible on a stable EC2 instance: https://github.com/WordPress/openverse-infrastructure/pull/829

Staci's early review comment made me realise I hadn't mentioned in the project thread that the security group refactor described in the implementation plan is not workable. Details of that [are described in this comment on the issue that was meant to implement the abstracted security group module](https://github.com/WordPress/openverse-infrastructure/issues/771#issuecomment-1970308850).

To summarise: the proposed abstraction in the implementation plan largely misses the point of security groups and how to best organise them. Rather than enforcing a uniform basic standard of ingress/egress rules by applying those rules to each security group individually, we should add instances to relevant security groups configured with the rules relevant to them. We can have several thousands of security groups, and with our service volume we will not run into the limit, even if we had an individual security group for each and every rule (which we don't need). Rather than configuring each EC2 instance's security group with SSH ingress rules, we should just add the instance to a shared security group with those rules.

That will be a long term refactor that I need to sit down and plan out into discrete tasks. We cannot do it as part of this project without causing significant delays to it as well as disruptions to virtually all other ongoing infrastructure work. I don't think this needs an implementation plan, it just needs someone (very likely me) to sit down and look at all our existing security groups to identify the places that need changes. Another issue that complicates this is the need to migrate away from both inline security group rules in Terraform _and_ away from the old security group rule resources, all towards [the new rule resources which have the significant benefit of leveraging AWS's relatively new security group rule ids](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_egress_rule), which neither of the previous approaches (which are the ones we use) implement.

So, we will still eventually change how we manage security groups to reduce duplication, we just won't do it the way the implementation plan suggests, and we will not do it as part of this project.

**Comment by sarayourfriend** [comment]
Airflow is live at airflow.openverse.org :tada: 

A final infrastructure PR is up to finalise the Ansible, compose, monitoring, and IAM configuration. Please review this as soon as possible, @WordPress/openverse-catalog @WordPress/openverse-infrastructure. The resources are live in production and ideally these changes are merged to `main` soon.

**Comment by sarayourfriend** [comment]
The API is now available at api.openverse.org :tada:

**Comment by sarayourfriend** [comment]
I've drafted text for the Make post and email https://docs.google.com/document/d/1ESmzbH6vkp8rxJBsy3P0_BLZ01sgKVFAQa3LnPSvBhQ/edit?usp=sharing

@WordPress/openverse-maintainers, please review the text. I'll start working on the baseline requirements for the management command in https://github.com/WordPress/openverse/issues/3742.

I'll shortly have a PR up for #3741.

**Comment by sarayourfriend** [comment]
I've got the make post up and scheduled, with a switch-over date of 3 June 2024, which @zackkrida and I just decided on. That gives a month of lead time, with the post scheduled to publish 6 May at 00:00 UTC.

This PR introduces the management command for sending the email to registered and verified API users: https://github.com/WordPress/openverse/pull/4229

Finally, https://github.com/WordPress/openverse-infrastructure/pull/876 updates the canonical URL and introduces a staging-only redirect for testing.

**Comment by sarayourfriend** [comment]
The Make post went out earlier this week: https://make.wordpress.org/openverse/2024/05/06/the-openverse-api-is-moving-to-api-openverse-org/

I got all set up to run the management command to send notifications to registered API users, but found an issue with the query in the original code, and it only pulled 2 email addresses in production to send to. That's not the expected outcome. I dug deeper, and realised we made a mistake in how we wrote the query. We'd written the query off of the `OAuth2Verification` model, thinking that it would be an easy way to find the email addresses of verified registrations. However, we delete the verification row once the email is verified, so that's not a sound approach.

Instead, we need to query off the registration table where `name` is the name of a verified application. This is safe because `name`  is unique indexed on the registration table.

I've confirmed that with this re-written query, we get a more expected number of emails to send in production.

**Comment by sarayourfriend** [comment]
The emails announcing the API move are sent as of 2024-05-08T00:59:58.366Z.

**Comment by sarayourfriend** [comment]
The redirects are **LIVE**! Everything appears to be working. I tried out Gutenberg and Jetpack integrations and both look to be fine as far as I can tell.

I've removed https://github.com/WordPress/openverse-infrastructure/issues/781 and https://github.com/WordPress/openverse-infrastructure/issues/782 from the milestones for this project because we agreed they should not block the "shipped" status of this project, and I wanted to make that clear based on the milestone.

With that, the API tickets are done, and I've closed the milestone!

I'm going to start working on https://github.com/WordPress/openverse-infrastructure/issues/785 now, which will free up https://github.com/WordPress/openverse-infrastructure/issues/787 very soon :tada:.

**Comment by zackkrida** [comment]
@sarayourfriend sadly, the redirect broke the `Inserter > Media > Openverse` flow in Gutenberg and (surprisingly to me) the Gutenberg E2E tests, which rely on availability of the Openverse API. There was some chat about this in the [Make WP Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1717419383604539) (public, but requires [an account](https://make.wordpress.org/chat/)).

I made an [infra PR](https://github.com/WordPress/openverse-infrastructure/pull/921) to revert the change, and a [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/62241) was already merged to replace the URLs to use openverse.org. 

To keep old versions of WordPress working, now, we would have to keep openverse.engineering operational indefinitely. That is clearly infeasible, so instead, I think we should re-implement the redirect right after WordPress 6.6 launches.

I also added a status update to our https://make.wordpress.org/openverse/2024/05/06/the-openverse-api-is-moving-to-api-openverse-org/ make post for the change, and pinned that post as we figure out next steps.

**Comment by sarayourfriend** [comment]
We've got a solution to redirect everything _except_ the media inserter. The solution doesn't prevent the goals of this project from succeeding as it only uses free Cloudflare features and can theoretically exist indefinitely. We can discuss when specifically we would remove it, whether that's with the 6.6 launch or later, in the follow up to this work.

**Comment by zackkrida** [comment]
Once https://github.com/WordPress/openverse-infrastructure/pull/920 is merged this project can be moved from shipped to success :partying_face:

**Comment by sarayourfriend** [comment]
I've just applied and merged https://github.com/WordPress/openverse-infrastructure/pull/920!

This project is complete :tada:
